src/engine/condition.ts

```diff
@@ -1,45 +1,51 @@
 import * as reducers from './reducers';
 import * as matchers from './matchers';
 import * as filters from './filters';
-import { PageView, EngineCondition } from '../../types';
-import { isNumberArray, isStringArray } from '../utils';
+import {
+  PageView,
+  PageFeatureResult,
+  EngineCondition,
+  EngineConditionQuery,
+  ArrayIntersectsFilter,
+  VectorDistanceFilter,
+  CosineSimilarityFilter,
+  AudienceDefinitionFilter,
+  VectorQueryValue,
+} from '../../types';
+import {
+  isNumberArray,
+  isStringArray,
+  isArrayIntersectsFilter,
+  isVectorDistanceFilter,
+  isCosineSimilarityFilter,
+} from '../utils';
 
-const createCondition = (condition: EngineCondition) => (
-  pageViews: PageView[]
-): boolean => {
+const createCondition = (condition: EngineCondition<AudienceDefinitionFilter>) =>
+  (pageViews: PageView[]): boolean => {
   const { filter, rules } = condition;
   const filteredPageViews = filter.queries
-    .map((query) => {
-      return pageViews.filter((pageView) => {
-        const queryFeatures = pageView.features[query.property];
-
-        if (query.filterComparisonType === 'arrayIntersects') {
-          return (
-            !!queryFeatures &&
-            queryFeatures.version === query.version &&
-            isStringArray(queryFeatures.value) &&
-            filters.arrayIntersects(queryFeatures.value, query.value)
-          );
-        } else if (query.filterComparisonType === 'vectorDistance') {
-          return (
-            !!queryFeatures &&
-            queryFeatures.version === query.version &&
-            isNumberArray(queryFeatures.value) &&
-            filters.vectorDistance(queryFeatures.value, query.value)
-          );
-        } else if (query.filterComparisonType === 'cosineSimilarity') {
-          return (
-            !!queryFeatures &&
-            queryFeatures.version === query.version &&
-            isNumberArray(queryFeatures.value) &&
-            filters.cosineSimilarity(queryFeatures.value, query.value)
-          );
-        } else {
-          return true;
-        }
-      });
-    })
-    .flat();
+  .map((query) => {
+    return pageViews.filter((pageView) => {
+      const queryFeatures: PageFeatureResult = pageView.features[query.queryProperty];
+      return isArrayIntersectsFilter(query)
+        ? isArrayIntersects(queryFeatures, query)
+        : isVectorDistanceFilter(query)
+        ? isVectorDistanceLesserThanThreshold(queryFeatures, query)
+        : isCosineSimilarityFilter(query)
+        ? isCosineSimilarityLesserThanThreshold(queryFeatures, query)
```

Comment:
I am not a fan of the nested ternary, it looks neat but its hard to follow, I am happy to leave it for now if you do not have any other ideas on how to handle this neatly. Quick one would be:

```
const filterType = checkQueryFilter(query);
filter(queryFeatures, query, filterType);
```
But I guess this could have issues with TS...

Also `isVectorDistanceLesserThanThreshold`, we match on greater than threshold.

Reply:
I agree. This is used to satisfy TS checker on the type specialization.

On your idea, I think, `checkQueryFilter` return type would be yet an union and so the `filter` function would not have been specialized.

I'll try to come up with a better solution.